client/client.go

```diff
@@ -0,0 +1,150 @@
+package client
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-delegated-routing/parser"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/multiformats/go-multiaddr"
+
+	//"github.com/ipld/go-ipld-prime/codec/dagjson"
+	"io"
+	"net/http"
+)
+
+type Option func(*client) error
+
+type client struct {
+	client *http.Client
+	endPoint string
+}
+
+func WithHTTPClient(hc *http.Client) Option {
+	return func(c *client) error {
+		c.client = hc
+		return nil
+	}
+}
+
+func New(endpoint string, opts...Option) (*client, error) {
+	c := &client{endPoint: endpoint, client: http.DefaultClient}
+	for _, o := range opts {
+		if err := o(c); err != nil {
+			return nil, err
+		}
+	}
+	return c, nil
+}
+
+func (c *client) FindProviders(ctx context.Context, cid cid.Cid) ([]peer.AddrInfo, error) {
+	req := parser.Envelope{
+		Tag: parser.MethodGetP2PProvide,
+		Payload: parser.GetP2PProvide{
+			Key: parser.ToDJSpecialBytes(cid.Hash()),
+		},
+	}
+	b := &bytes.Buffer{}
+	if err := json.NewEncoder(b).Encode(req); err != nil {
+		return nil, err
+	}
+
+	httpReq, err := http.NewRequestWithContext(ctx, "POST", c.endPoint, b)
```

Comment:
Is there a reason for "POST"? This will break HTTP middleware caches.

Reply:
How were you envisioning GET here, Base64URL encoding the JSON and putting it in the query string or something? This doesn't seem like the thing to do (e.g. https://stackoverflow.com/questions/6073896/sending-json-object-to-the-server-via-http-get/6073954).

Reply:
for 'get CID' requests it's an immutable record and should be a GET.
* In the backup handler nitro want to be able to fall back to it would need to be.
* we would prefer sticking closer to natural protocol semantics / rest here, which would make this a get

Reply:
> for 'get CID' requests it's an immutable record

How? You can receive new advertisements (or unadvertisements) over time which would modify the return value.

Reply:
you're right! it probably should be cache-able, but we'll need to figure out a reasonable update/expiration

Reply:
> we would prefer sticking closer to natural protocol semantics / rest here

As per previous discussions we want streaming responses. We have two implementations of content routing at the moment, one streams responses and the other one does not. Having a generic protocol that doesn't support one out of the two existing systems is just not generic.

Streaming responses from GET requests is not really "REST" semantics.

This protocol is closer to JSON-RPC in that way (we could even make it fully JSON-RPC if people really wanted to although AFAICT benefits don't seem very meaningful especially if/when we allow for nested parameters in methods like authentication, signatures or bulk methods).

There are a variety of resources online indicating that GET is not preferred for things like JSON-RPC, such as https://www.simple-is-better.org/json-rpc/transport_http.html#get-request.

That being said if using middleware with HTTP caching of GET requests is really that important we can probably just extend the `identify` method to help clients figure out which methods are safe for GET requests.

Reply:
you are `GET`'ing rather than mutating any data in the request. All mechanisms for streaming will be the same regardless of the HTTP verb used - you can trickle back responses line-by-line from the server in the same way you would in responding to a post request, and the golang client will not handle things any differently based on the verb.